service/src/main/java/bio/terra/cbas/initialization/CloneRecoveryService.java

```diff
@@ -0,0 +1,96 @@
+package bio.terra.cbas.initialization;
+
+import bio.terra.cbas.config.CbasContextConfiguration;
+import bio.terra.cbas.dao.MethodDao;
+import bio.terra.cbas.dao.RunDao;
+import bio.terra.cbas.dao.RunSetDao;
+import bio.terra.cbas.models.Method;
+import bio.terra.cbas.models.Run;
+import bio.terra.cbas.models.RunSet;
+import bio.terra.common.db.WriteTransaction;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CloneRecoveryService {
+
+  private final Logger logger = LoggerFactory.getLogger(CloneRecoveryService.class);
+  private final RunDao runDao;
+  private final RunSetDao runSetDao;
+  private final MethodDao methodDao;
+  private final CbasContextConfiguration cbasContextConfig;
+
+  public CloneRecoveryService(
+      RunSetDao runSetDao,
+      RunDao runDao,
+      MethodDao methodDao,
+      CbasContextConfiguration cbasContextConfig) {
+    this.runSetDao = runSetDao;
+    this.runDao = runDao;
+    this.methodDao = methodDao;
+    this.cbasContextConfig = cbasContextConfig;
+  }
+
+  public void cloneRecovery() {
+    logger.info(
+        "Starting clone recovery (workspaceId: {}, workspaceCreatedDate: {})",
+        cbasContextConfig.getWorkspaceId(),
+        cbasContextConfig.getWorkspaceCreatedDate());
+
+    pruneCloneSourceWorkspaceHistory();
+  }
+
+  @WriteTransaction
+  public void pruneCloneSourceWorkspaceHistory() {
+    List<Method> methods = methodDao.getMethods().stream().filter(this::isMethodCloned).toList();
+
+    methods.forEach(this::updateMethodTemplate);
+
+    methods.forEach(this::cleanupClonedMethodHistory);
+  }
+
+  public Boolean isMethodCloned(Method m) {
+    return !m.originalWorkspaceId().equals(cbasContextConfig.getWorkspaceId());
+  }
+
+  public Boolean isRunSetCloned(RunSet r) {
+    return !r.originalWorkspaceId().equals(cbasContextConfig.getWorkspaceId());
+  }
+
+  @WriteTransaction
```

Comment:
I believe transactions are re-entrant rather than nested, so this will probably just be part of the same original transaction as you created in the caller... except that it's called from the same class so would be ignored anyway.